File: .opencode/tools/tabphm_sanity/tabphm_sanity.py
"""CLI entrypoint for PICID sanity tools.

OpenCode custom tools call this script and pass a JSON payload through
``--input-json``. The script prints one marker-prefixed JSON result so callers can
ignore framework logging that may appear on stdout.
"""

from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import Any

from PICID_sanity_core import (
    DATASET_EXECUTION_FAILED,
    DATASET_UNAVAILABLE,
    PRECHECK_TIMEOUT,
    TOOL_INVOCATION_FAILURE,
    CheckResult,
    append_report,
    check_gradient_flow,
    check_init_loss,
    check_overfit_batch,
    check_subset_convergence,
    check_zero_input,
    ensure_project_root,
    json_default,
    ladder_result,
    preflight_config,
    render_markdown,
    skip_gradient_check,
)


RESULT_MARKER = "PICID_SANITY_RESULT_JSON="


def payload_arg(payload: dict[str, Any], key: str, default: Any = None) -> Any:
    value = payload.get(key, default)
    return default if value is None else value


def common(payload: dict[str, Any]) -> dict[str, Any]:
    repo_root = ensure_project_root(payload_arg(payload, "repo_root", "."))
    return {
        "repo_root": repo_root,
        "experiment_config": payload_arg(payload, "experiment_config"),
        "vault_dir": payload.get("vault_dir"),
        "task_type": payload_arg(payload, "task_type", "regression"),
        "dataset_candidate": payload.get("dataset_candidate"),
        "extra_overrides": list(payload_arg(payload, "extra_overrides", [])),
        "timeout_sec": payload.get("timeout_sec"),
        "write_log": bool(payload_arg(payload, "write_log", False)),
        "zero_keys": payload.get("zero_keys"),
        "micro_batch_size": int(payload_arg(payload, "micro_batch_size", 4)),
        "overfit_max_steps": int(payload_arg(payload, "overfit_max_steps", 400)),
        "include_subset_convergence": bool(
            payload_arg(payload, "include_subset_convergence", False)
        ),
    }


def finish(result: dict[str, Any]) -> None:
    print(RESULT_MARKER + json.dumps(result, default=json_default, sort_keys=True))


def single_result(
    check: CheckResult,
    payload: dict[str, Any],
    values: dict[str, Any],
) -> dict[str, Any]:
    result = ladder_result(
        checks=[check],
        experiment_config=values["experiment_config"],
        task_type=values["task_type"],
        dataset_candidate=values["dataset_candidate"],
        vault_dir=values["vault_dir"],
        write_log=values["write_log"],
        compact=True,
    )
    result["tool"] = payload.get("command")
    return result


def run_command(command: str, payload: dict[str, Any]) -> dict[str, Any]:
    values = common(payload)
    experiment_config = values["experiment_config"]
    if not experiment_config:
        raise ValueError("experiment_config is required")

    task_type = values["task_type"]
    if command == "init_loss":
        check = check_init_loss(
            repo_root=values["repo_root"],
            experiment_config=experiment_config,
            task_type=task_type,
            num_classes=payload.get("num_classes"),
            expected_init_loss=payload.get("expected_init_loss", "auto"),
            extra_overrides=values["extra_overrides"],
            timeout_sec=values["timeout_sec"],
        )
        return single_result(check, payload, values)

    if command == "zero_input":
        if task_type == "fit_predict":
            check = CheckResult(
                name="zero_input",
                status="SKIPPED",
                diagnostic="fit-predict zero-input check is not implemented by the tool-side manual trainer",
            )
        else:
            check = check_zero_input(
                repo_root=values["repo_root"],
                experiment_config=experiment_config,
                extra_overrides=values["extra_overrides"],
                zero_keys=values["zero_keys"],
                timeout_sec=values["timeout_sec"],
            )
        return single_result(check, payload, values)

    if command == "overfit_batch":
        if task_type == "fit_predict":
            check = skip_gradient_check("overfit_batch")
        else:
            check = check_overfit_batch(
                repo_root=values["repo_root"],
                experiment_config=experiment_config,
                task_type=task_type,
                extra_overrides=values["extra_overrides"],
                timeout_sec=values["timeout_sec"],
                micro_batch_size=values["micro_batch_size"],
                max_steps=values["overfit_max_steps"],
            )
        return single_result(check, payload, values)

    if command == "gradient_flow":
        if task_type == "fit_predict":
            check = skip_gradient_check("gradient_flow")
        else:
            check = check_gradient_flow(
                repo_root=values["repo_root"],
                experiment_config=experiment_config,
                extra_overrides=values["extra_overrides"],
                zero_keys=values["zero_keys"],
                timeout_sec=values["timeout_sec"],
            )
        return single_result(check, payload, values)

    if command == "subset_convergence":
        if task_type == "fit_predict":
            check = skip_gradient_check("subset_convergence")
        else:
            check = check_subset_convergence(
                repo_root=values["repo_root"],
                experiment_config=experiment_config,
                extra_overrides=values["extra_overrides"],
                timeout_sec=values["timeout_sec"],
            )
        return single_result(check, payload, values)

    if command == "ladder":
        checks: list[CheckResult] = []
        preflight = preflight_config(
            values["repo_root"], experiment_config, values["extra_overrides"]
        )
        checks.append(preflight)
        if preflight.status != "PASS":
            return ladder_result(
                checks=checks,
                experiment_config=experiment_config,
                task_type=task_type,
                dataset_candidate=values["dataset_candidate"],
                vault_dir=values["vault_dir"],
                write_log=bool(payload_arg(payload, "write_log", True)),
                compact=True,
            )
        if task_type == "fit_predict":
            checks.extend(
                [
                    skip_gradient_check("init_loss"),
                    skip_gradient_check("gradient_flow"),
                    skip_gradient_check("overfit_batch"),
                ]
            )
            if values["include_subset_convergence"]:
                checks.append(skip_gradient_check("subset_convergence"))
        else:
            ladder_steps = [
                lambda: check_init_loss(
                    repo_root=values["repo_root"],
                    experiment_config=experiment_config,
                    task_type=task_type,
                    num_classes=payload.get("num_classes"),
                    expected_init_loss=payload.get("expected_init_loss", "auto"),
                    extra_overrides=values["extra_overrides"],
                    timeout_sec=values["timeout_sec"],
                ),
                lambda: check_gradient_flow(
                    repo_root=values["repo_root"],
                    experiment_config=experiment_config,
                    extra_overrides=values["extra_overrides"],
                    zero_keys=values["zero_keys"],
                    timeout_sec=values["timeout_sec"],
                ),
                lambda: check_overfit_batch(
                    repo_root=values["repo_root"],
                    experiment_config=experiment_config,
                    task_type=task_type,
                    extra_overrides=values["extra_overrides"],
                    timeout_sec=values["timeout_sec"],
                    micro_batch_size=values["micro_batch_size"],
                    max_steps=values["overfit_max_steps"],
                ),
            ]
            if values["include_subset_convergence"]:
                ladder_steps.append(
                    lambda: check_subset_convergence(
                        repo_root=values["repo_root"],
                        experiment_config=experiment_config,
                        extra_overrides=values["extra_overrides"],
                        timeout_sec=values["timeout_sec"],
                    )
                )
            for check_fn in ladder_steps:
                check = check_fn()
                checks.append(check)
                if check.status in (
                    DATASET_UNAVAILABLE,
                    DATASET_EXECUTION_FAILED,
                    TOOL_INVOCATION_FAILURE,
                    PRECHECK_TIMEOUT,
                ):
                    break
        return ladder_result(
            checks=checks,
            experiment_config=experiment_config,
            task_type=task_type,
            dataset_candidate=values["dataset_candidate"],
            vault_dir=values["vault_dir"],
            write_log=bool(payload_arg(payload, "write_log", True)),
            compact=True,
        )

    if command == "report":
        checks = [CheckResult(**item) for item in payload_arg(payload, "checks", [])]
        result = ladder_result(
            checks=checks,
            experiment_config=experiment_config,
            task_type=task_type,
            dataset_candidate=values["dataset_candidate"],
            vault_dir=values["vault_dir"],
            write_log=False,
        )
        markdown = render_markdown(
            result,
            values["vault_dir"],
            experiment_config,
            task_type,
            values["dataset_candidate"],
        )
        result["markdown_appended_path"] = append_report(markdown, values["vault_dir"])
        return result

    raise ValueError(f"Unknown command: {command}")


def main() -> None:
    parser = argparse.ArgumentParser(description="Run PICID sanity checks.")
    parser.add_argument(
        "command",
        choices=[
            "ladder",
            "init_loss",
            "zero_input",
            "overfit_batch",
            "gradient_flow",
            "subset_convergence",
            "report",
        ],
    )
    parser.add_argument("--input-json", required=True)
    args = parser.parse_args()
    payload = json.loads(args.input_json)
    payload["command"] = args.command
    result = run_command(args.command, payload)
    finish(result)


if __name__ == "__main__":
    main()

File: .opencode/tools/tabphm_sanity.ts
import { tool, type ToolContext } from "@opencode-ai/plugin"
import path from "path"

const z = tool.schema
const RESULT_MARKER = "PICID_SANITY_RESULT_JSON="

const commonArgs = {
  experiment_config: z.string().describe("Hydra experiment config, e.g. pronostia/prognostics/raw/my_model"),
  vault_dir: z.string().optional().describe("Vault directory containing 04-implementation-blueprint.md"),
  repo_root: z.string().optional().describe("Repository root. Defaults to the OpenCode worktree."),
  dataset_candidate: z.string().optional().describe("Current dataset candidate name for fallback reporting"),
  task_type: z.enum(["regression", "classification", "fit_predict"]).default("regression"),
  num_classes: z.number().optional().describe("Number of classes for classification init-loss checks"),
  expected_init_loss: z.union([z.number(), z.literal("auto")]).default("auto"),
  extra_overrides: z.array(z.string()).optional().describe("Additional Hydra overrides appended to sanity commands"),
  timeout_sec: z.number().optional().describe("Per-check wall-clock timeout in seconds (default 600s for in-process checks)"),
  write_log: z.boolean().optional().describe("Append Markdown to vault_dir/06-sanity-ladder-log.md"),
  zero_keys: z.array(z.string()).optional().describe("Batch keys to zero for input-independence and gradient checks"),
  micro_batch_size: z.number().optional().describe("Sample count for the micro-batch memorization probe used by overfit_batch (default 4)"),
  overfit_max_steps: z.number().optional().describe("Maximum optimizer steps for the micro-batch memorization probe (default 400)"),
  include_subset_convergence: z.boolean().optional().describe("Opt in to the expensive 30-epoch subset_convergence check (off by default)"),
}

async function invoke(command: string, args: Record<string, unknown>, context: ToolContext) {
  const repoRoot = String(args.repo_root || context.worktree || context.directory)
  const script = path.join(repoRoot, ".opencode", "tools", "PICID_sanity", "PICID_sanity.py")
  const payload = JSON.stringify({ ...args, repo_root: repoRoot })
  const proc = Bun.spawn(["uv", "run", "python", script, command, "--input-json", payload], {
    cwd: repoRoot,
    stdout: "pipe",
    stderr: "inherit",
    env: {
      ...process.env,
      PROJECT_ROOT: repoRoot,
      TABPFN_DISABLE_TELEMETRY: "1",
    },
  })
  const stdout = await new Response(proc.stdout).text()
  const exitCode = await proc.exited
  if (exitCode !== 0) {
    throw new Error(`PICID sanity tool failed with exit code ${exitCode}\nSTDOUT:\n${stdout}`)
  }
  const markerIndex = stdout.lastIndexOf(RESULT_MARKER)
  if (markerIndex < 0) {
    throw new Error(`PICID sanity tool did not emit ${RESULT_MARKER}\nSTDOUT:\n${stdout}`)
  }
  const raw = stdout.slice(markerIndex + RESULT_MARKER.length).trim()
  const parsed = JSON.parse(raw)
  return JSON.stringify(parsed, null, 2)
}

export const ladder = tool({
  description: "Run the trimmed 3-check PICID sanity ladder (init_loss → gradient_flow → overfit_batch) in a single in-process framework build, with [heartbeat] progress on stderr and a 600s default per-check timeout. The overfit_batch step is a micro-batch memorization probe. Set include_subset_convergence:true only when explicitly needed.",
  args: commonArgs,
  async execute(args, context) {
    return invoke("ladder", { ...args, write_log: args.write_log ?? true }, context)
  },
})

export const init_loss = tool({
  description: "Run PICID sanity Check 1: loss at initialization.",
  args: commonArgs,
  async execute(args, context) {
    return invoke("init_loss", args, context)
  },
})

export const zero_input = tool({
  description: "Run PICID sanity Check 2: real inputs versus zeroed inputs.",
  args: commonArgs,
  async execute(args, context) {
    return invoke("zero_input", args, context)
  },
})

export const overfit_batch = tool({
  description: "Run PICID sanity Check 3: memorize a tiny sliced micro-batch and classify the result as PASS / INVESTIGATE / FAIL.",
  args: commonArgs,
  async execute(args, context) {
    return invoke("overfit_batch", args, context)
  },
})

export const gradient_flow = tool({
  description: "Run PICID sanity Check 4: gradient-flow and batch-isolation audit.",
  args: commonArgs,
  async execute(args, context) {
    return invoke("gradient_flow", args, context)
  },
})

export const subset_convergence = tool({
  description: "Run PICID sanity small-subset convergence check (opt-in only; expensive 30-epoch CLI run, not part of the default ladder).",
  args: commonArgs,
  async execute(args, context) {
    return invoke("subset_convergence", args, context)
  },
})

export const report = tool({
  description: "Append a PICID sanity Markdown report from structured check results.",
  args: {
    ...commonArgs,
    checks: z.array(z.any()).describe("Check result objects returned by other PICID_sanity tools"),
  },
  async execute(args, context) {
    return invoke("report", args, context)
  },
})
